src/NServiceBus.Core/OpenTelemetry/Metrics/MessageHandlingMetricsFactory.cs

```diff
@@ -0,0 +1,50 @@
+﻿namespace NServiceBus;
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Pipeline;
+
+class MessageHandlingMetricsFactory(string queueName, string discriminator) : IMessageHandlingMetricsFactory
+{
+    public IMessageHandlingMetrics StartHandling(IInvokeHandlerContext context)
+    {
+        var messageType = context.MessageBeingHandled?.GetType().FullName;
+        var tagList = new TagList(new KeyValuePair<string, object>[]
+        {
+            new(MeterTags.QueueName, queueName ?? ""),
+            new(MeterTags.EndpointDiscriminator, discriminator ?? ""),
+            new(MeterTags.MessageType, messageType ?? "")
+        }.AsSpan());
+        var handlerType = context.MessageHandler.Instance?.GetType().FullName;
+        tagList.Add(MeterTags.MessageHandlerType, handlerType ?? "");
```

Comment:
Given we're always adding handler type, shouldn't we reorder and construct in one shot?
```suggestion
        var handlerType = context.MessageHandler.Instance?.GetType().FullName;
        var tagList = new TagList(new KeyValuePair<string, object>[]
        {
            new(MeterTags.QueueName, queueName ?? ""),
            new(MeterTags.EndpointDiscriminator, discriminator ?? ""),
            new(MeterTags.MessageType, messageType ?? ""),
            new(MeterTags.MessageHandlerType, handlerType ?? "")
        }.AsSpan());
```